File: backend/app/schemas/resource.py
"""资源推荐与绑定 schemas"""
from __future__ import annotations

from datetime import datetime
from urllib.parse import urlparse

from pydantic import BaseModel, Field, field_validator, model_validator


class ResourceItem(BaseModel):
    id: str
    title: str
    url: str
    snippet: str | None = None
    score: float | None = None
    source_type: str
    stage_name: str | None = None
    node_id: str | None = None
    preference_match: str | None = None
    preference_reason: str | None = None
    created_at: datetime | None = None


class NodeResourceGroup(BaseModel):
    node_id: str
    node_name: str
    resources: list[ResourceItem] = Field(default_factory=list)


class StageResourceGroup(BaseModel):
    stage_name: str
    stage_resources: list[ResourceItem] = Field(default_factory=list)
    nodes: list[NodeResourceGroup] = Field(default_factory=list)


class PlanResourcesResponse(BaseModel):
    path_id: str
    stages: list[StageResourceGroup]


class ManualResourceBindRequest(BaseModel):
    stage_name: str | None = None
    node_id: str | None = None
    title: str = Field(min_length=1)
    url: str = Field(min_length=1)
    snippet: str | None = None

    @field_validator("url")
    @classmethod
    def validate_url(cls, value: str) -> str:
        normalized = value.strip()
        parsed = urlparse(normalized)
        if parsed.scheme not in {"http", "https"} or not parsed.netloc:
            raise ValueError("url 仅支持 http/https 链接")
        return normalized

    @model_validator(mode="after")
    def validate_target(self) -> "ManualResourceBindRequest":
        if not self.stage_name and not self.node_id:
            raise ValueError("stage_name 和 node_id 至少提供一个")
        return self

File: frontend/src/api/modules/resource.ts
import request from '../request'

export interface ResourceItem {
  id: string
  title: string
  url: string
  snippet?: string | null
  score?: number | null
  source_type: 'static' | 'tavily_auto' | 'manual' | string
  preference_match?: 'mixed' | 'preferred' | 'available' | string | null
  preference_reason?: string | null
  stage_name?: string | null
  node_id?: string | null
  created_at?: string | null
}

export interface NodeResourceGroup {
  node_id: string
  node_name: string
  resources: ResourceItem[]
}

export interface StageResourceGroup {
  stage_name: string
  stage_resources: ResourceItem[]
  nodes: NodeResourceGroup[]
}

export interface PlanResourcesResponse {
  path_id: string
  stages: StageResourceGroup[]
}

export interface ProjectResourceBindingRequest {
  resource_id: string
  target_type: 'project_node' | 'path_stage'
  target_id: string
  source_result_id?: string | null
  binding_source?: string
}

export interface ProjectResourceBindingResponse {
  id: string
  project_id: string
  resource_id: string
  target_type: string
  target_id: string
  source_result_id?: string | null
  binding_source: string
  created_at: string
}

export const resourceApi = {
  getPlanResources: (projectId: string, pathId: string): Promise<PlanResourcesResponse> =>
    request.get(`/projects/${projectId}/plans/${pathId}/resources`),
  recommendPlanResources: (projectId: string, pathId: string): Promise<PlanResourcesResponse> =>
    request.post(`/projects/${projectId}/plans/${pathId}/resources/recommend`),
  bindManualResource: (
    projectId: string,
    pathId: string,
    payload: { stage_name?: string; node_id?: string; title: string; url: string; snippet?: string }
  ): Promise<ResourceItem> =>
    request.post(`/projects/${projectId}/plans/${pathId}/resources/bind`, payload),
  bindProjectResource: (
    projectId: string,
    payload: ProjectResourceBindingRequest,
  ): Promise<ProjectResourceBindingResponse> =>
    request.post(`/projects/${projectId}/resources/bindings`, payload),
}
